### ISSUE TITLE
tests with RefreshDatabase not working after schema:dump --prune

### ISSUE THREAD

**Issue by retosteffen** [laravel/framework#36294]
- Laravel Version: 8.28.1
- PHP Version: 7.4.13
- Database Driver & Version: mysql 5.7.32

### Description:
I cleaned up my migrations with schema:dump --prune but then RefreshDatabase doesn't correctly migrate my test database anymore.
If I set up my test with an explicit schema path it works $this->artisan('migrate:fresh --schema-path "database/schema/mysql-schema.dump"');

**Comment by taylorotwell** [comment]
Unable to recreate this.

**Comment by tombanksme** [comment]
I've encountered this issue and think I've managed to reproduce it in a fresh Laravel installation.
- Laravel Version: 8.30.0
- PHP Version: 8.0.2
- MySQL Version: 8.0.23-0ubuntu0.20.04.1

I'll link a repository containing my code below but here is an outline of what I did:
- Created a new Laravel project.
- Created a "testing" connection in my config/database.php file & updated the phpunit.xml file to use it.
- Added RefreshDatabase trait to my TestCase class
- Created the Purchase model.
- Migrated, dumped & pruned the database.
- In the Feature/ExampleTest I attempt to get all Purchases from the database and get a table not found error.

Hope this helps. https://github.com/tombanksme/laravel-prune-testing-issue

**Comment by flexchar** [comment]
@tombanksme you may need to recreate it. It seems like they are very hard on closing the issues. :)

**Comment by moff** [comment]
**For all of those who struggled with running tests against database using Laravel 8 dump file**

As it turned out after hours of head banging against the wall, if you want to run migrations using schema dump in test environment you have these possible solutions:

- create schema dump named accordingly to the test connection specified in `config/database.php` (e.g. if you have `pgsql_testing` connection, you name the schema dump as `pgsql_testing-schema.dump`)
- run migrations manually before running `artisan test` using the key `--schema-path` to use dump created by standard command (e.g. in my case it was `--schema-path=database/schema/pgsql-schema.dump`)

To create dump I used command `php artisan schema:dump --prune` mentioned in documentation and got dump file `pgsql-schema.dump` in `database/schema`

So basically it's all about the fact that during running tests and using migrations dump it won't be used unless the dump file properly named. 

@taylorotwell that cost me hours of my time, it would be wise to put that in documentation, because it's totally implicit behaviour. I don't work much with Laravel so maybe it's obvious for others, but that wasn't for me.

**Comment by nsbucky** [comment]
when creating a schema dump, use the name of the connection in the database config array, NOT the name of your actual database. Most likely if you are using default setup with a key of "testing" in the database.php file, create/copy your .dump file to: /database/schema/testing-schema.dump

Your schema will get picked up during tests, you will not have to run migrations manually at all.

**Comment by erikverbeek** [comment]
@allanvobraun building on your workaround I created:

```php
abstract class TestCase extends BaseTestCase
{
    use CreatesApplication, RefreshDatabase {
        migrateFreshUsing as traitMigrateFreshUsing;
    }

    protected function migrateFreshUsing()
    {
        return array_merge(
            [
                '--schema-path' => 'database/schema/mysql-schema.dump'
            ],
            $this->traitMigrateFreshUsing()
        );
    }
}
```

This way you won't loose the flags, `--drop-views`, `--drop-types` and `--seed`, that are set by the original `migrateFreshUsing` function.

**Comment by robjensen82** [comment]
Another approach is to define your own trait for refresh database:

```
trait MyRefreshDatabase {
    use RefreshDatabase {
        migrateFreshUsing as private originalMigrateFreshUsing;
    }

    protected function migrateFreshUsing(): array
    {
        return array_merge(
            [
                '--schema-path' => 'database/schema/pgsql-schema.dump'
            ],
            $this->originalMigrateFreshUsing()
        );
    }
}
```

**Comment by jasperf** [comment]
@robjensen82  I added

```php
<?php

namespace App\Http\Traits;

use Illuminate\Foundation\Testing\RefreshDatabase;

trait MyRefreshDatabase {

    use RefreshDatabase {
    migrateFreshUsing as private originalMigrateFreshUsing;
    }

    protected function migrateFreshUsing(): array
    {
        return array_merge(
            [
                '--schema-path' => 'database/schema/test-schema.dump'
            ],
            $this->originalMigrateFreshUsing()
        );
    }
}
```

but it does not seem to work using default sqlite as database connection. It still uses the default connection instead. When I used a separate test connection using mysql it did work but it was rather slow.

**Comment by robjensen82** [comment]
@jasperf Are you somehow running things twice? As in haven't deleted your migrations? Not sure why it would be slower.

**Comment by jasperf** [comment]
@robjensen82 Well, I have a JSON, loginController and registerController test containing 7 tests in total and it took 16 seconds to run all of these using a MySQL database where migrations had already run:

```
php artisan test

   PASS  Tests\Unit\JsonObjectTest
  ✓ get exists property
  ✓ get undefined property

   PASS  Tests\Feature\Http\Controllers\Auth\LoginControllerTest
  ✓ login displays the login form
  ✓ user cannot view a login form when authenticated
  ✓ login displays validation errors

   PASS  Tests\Feature\Http\Controllers\Auth\RegisterControllerTest
  ✓ registration request works
  ✓ register displays validation errors

  Tests:  7 passed
  Time:   16.48s
```

Just wondered if that was fast or slow and if so why slow. I expected it to be done in like 5 seconds.

I did not delete the migrations earlier. I think it just still used the wrong database connection, but now with
```html
<?xml version="1.0" encoding="UTF-8"?>
<phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:noNamespaceSchemaLocation="./vendor/phpunit/phpunit/phpunit.xsd"
         bootstrap="vendor/autoload.php"
         colors="true"
>
    <testsuites>
        <testsuite name="Unit">
            <directory suffix="Test.php">./tests/Unit</directory>
        </testsuite>
        <testsuite name="Feature">
            <directory suffix="Test.php">./tests/Feature</directory>
        </testsuite>
    </testsuites>
    <coverage processUncoveredFiles="true">
        <include>
            <directory suffix=".php">./app</directory>
        </include>
    </coverage>
    <php>
        <env name="APP_ENV" value="testing"/>
        <env name="BCRYPT_ROUNDS" value="4"/>
        <env name="CACHE_DRIVER" value="array"/>
        <env name="DB_CONNECTION" value="mysql"/>
        <env name="DB_DATABASE" value="test"/>
        <env name="MAIL_MAILER" value="array"/>
        <env name="QUEUE_CONNECTION" value="sync"/>
        <env name="SESSION_DRIVER" value="array"/>
        <env name="TELESCOPE_ENABLED" value="false"/>
    </php>
</phpunit>
```

it does work.

**Comment by JadoJodo** [comment]
For those who find this issue via a search engine:

I had an issue that manifested similarly to this (and subsequently found this issue), but it did not end up being this exact issue. What it did end up being:

A migration that added a default row to a table (to ensure consistent values for models that rely on that table) was dropped during the schema dump (by design). But because it was also not a part of the DB schema (by design), it disappeared. As a result, tests that relied on that model's existence started failing.
